src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -0,0 +1,161 @@
+        // iterate over results from each shard. Every CompoundTopDocs object has results from
+        // multiple sub queries, doc ids may repeat for each sub query results
+        return queryTopDocs.stream()
+            .map(compoundQueryTopDocs -> combineShardScores(scoreCombinationTechnique, compoundQueryTopDocs))
+            .collect(Collectors.toList());
+    }
+
+    private float combineShardScores(ScoreCombinationTechnique scoreCombinationTechnique, CompoundTopDocs compoundQueryTopDocs) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+            return ZERO_SCORE;
+        }
+        List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+        // - create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(topDocsPerSubQuery);
+
+        // - create map of combined scores per doc id
+        Map<Integer, Float> combinedNormalizedScoresByDocId = combineScoresAndGetCombinedNormilizedScoresPerDocument(
+            normalizedScoresPerDoc,
+            scoreCombinationTechnique
+        );
+
+        // - sort documents by scores and take first "max number" of docs
+        // create a priority queue of doc ids that are sorted by their combined scores
+        PriorityQueue<Integer> scoreQueue = getPriorityQueueOfDocIds(normalizedScoresPerDoc, combinedNormalizedScoresByDocId);
+        // store max score to resulting list, call it now as priority queue will change after combining scores
+        float maxScore = combinedNormalizedScoresByDocId.get(scoreQueue.peek());
+
+        // - update query search results with normalized scores
+        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, scoreQueue);
+        return maxScore;
+    }
+
+    private PriorityQueue<Integer> getPriorityQueueOfDocIds(
+        Map<Integer, float[]> normalizedScoresPerDoc,
+        Map<Integer, Float> combinedNormalizedScoresByDocId
+    ) {
+        PriorityQueue<Integer> pq = new PriorityQueue<>(
+            (a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a))
+        );
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        pq.addAll(normalizedScoresPerDoc.keySet());
+        return pq;
+    }
+
+    private ScoreDoc[] getCombinedScoreDocs(
+        final CompoundTopDocs compoundQueryTopDocs,
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final PriorityQueue<Integer> scoreQueue,
+        final int maxHits
+    ) {
+        ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
+
+        int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+        for (int j = 0; j < maxHits; j++) {
+            int docId = scoreQueue.poll();
```

Comment:
We should consider to use sorted list instead of priority queue. The time complexity of the `poll()` operation is `log(n)`

Reply:
I'm not sure time complexity is entirely correct, for ```poll()``` it's ```O(1)```. ```PriorityQueue``` and it's subclasses are used all over OpenSearch and Lucene in similar cases.

Reply:
It depends on how the priority queue is implemented. However, there is a high chance that current implementation is O(logn). https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/PriorityQueue.html

Reply:
Looks like there are no clear answer, some sources say it's O(1) and some that it's O(logn). It's not super critical section, I'll try to rewrite it with a sorted list, otherwise we can leave it for performance test time.